app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Field.java

```diff
@@ -0,0 +1,339 @@
+package com.mithridat.nonoconverter.backend.nonogram;
+
+import android.graphics.Bitmap;
+import android.os.Parcel;
+import android.os.Parcelable;
+
+import java.util.Arrays;
+
+/**
+ * Field storage class
+ */
+public class Field implements Parcelable {
+
+    /**
+     * White color
+     */
+    public static final int WHITE = 0xFFFFFFFF;
+
+    /**
+     * Black color
+     */
+    public static final int BLACK = 0xFF000000;
+
+    /**
+     * Empty cell
+     */
+    public static final int EMPTY = -1;
+
+    /**
+     * Constants for argument type - row or column
+     */
+    public static final int ROW = 0, COL = 1;
+
+    /**
+     * Constants for state type - indexes or colors
+     */
+    public static final int INDEX = 0, COLOR = 1;
+
+    /**
+     * Field сolor matrix
+     */
+    private int[][] _field;
+
+    /**
+     * Field sizes
+     */
+    private int _rows, _cols;
+
+    /**
+     * Array of colors that used in a puzzle. All cells of the field
+     * contain values of colors. But now it will contain
+     * indexes of these colors in the array _colors by default.
+     */
+    private int[] _colors;
+
+    /**
+     * Length of the array _colors
+     */
+    private int _colorsCount;
+
+    /**
+     * Identifier that indicates what is stored in the field's cells.
+     * Both colors and indexes of these colors can be stored in the cells
+     */
+    private int _state;
+
+    /**
+     * Constructor by black-and-white image
+     *
+     * @param bmp - black-and-white image
+     * @param rows - number of thumbnail rows
+     * @param cols - number of thumbnail columns
+     * @param p - fill parameter
+     */
+    public Field(Bitmap bmp, int rows, int cols, int p) {
+        _colors = new int[]{ BLACK, WHITE };
+        _colorsCount = _colors.length;
+        _state = INDEX;
+        _rows = rows;
+        _cols = cols;
+        _field = new int[rows][cols];
+        int height = bmp.getHeight(), width = bmp.getWidth();
+        int h = height / rows, w = width / cols;
+        int sumColors;
+        for (int i = 0; i < rows; i++) {
+            for (int j = 0; j < cols; j++) {
+                sumColors = 0;
+                for (int x = j * w; x < (j + 1) * w; x++) {
+                    for (int y = i * h; y < (i + 1) * h; y++) {
+                        if (bmp.getPixel(x, y) == WHITE) sumColors += 255;
+                    }
+                }
+                if (sumColors <= p * w * h) {
+                    _field[i][j] = Arrays.binarySearch(_colors, BLACK);
+                }
+            }
+        }
+    }
+
+    public Field(Field other) {
```

Comment:
This method usually calls `clone` in Java.

Reply:
Sorry, the semantic of the `clone` is some another. You can rename it to `copy` if you want.